### PR TITLE
AAQ-385: Add pg vector

### PR DESCRIPTION
**Pull request by lickem22**
Reviewer: @suzinyou 

Estimate: 2h

---

## Ticket

Fixes: [AAQ-348](https://idinsight.atlassian.net/browse/AAQ-348?atlOrigin=eyJpIjoiY2Y0ZmViM2NiYzI3NDI1YmE4OWRjOTNmNDRjNWYyOTciLCJwIjoiaiJ9)
### Goal
The goal of this PR is to merge the pg vector feature to main. All changes have been reviewed before. So this PR is just to make sure everything works before merging to main. 
### Changes
All changes have been reviewed before. No additional changes have been made. 
## How has this been tested?
Doubled checked that everything worked and ran tests again
## Checklist

Fill with `x` for completed. Delete any lines that are not relevant

- [x] My code follows the style guidelines of this project
- [x] I have reviewed my own code to ensure good quality
- [x] I have tested the functionality of my code to ensure it works as intended
- [x] I have resolved merge conflicts
- [x] I have updated the automated tests (if applicable)
- [x] I have updated the requirements (if applicable)
- [x] I have updated the README file (if applicable)
- [x] I have updated affected documentation (if applicable)

[AAQ-348]: https://idinsight.atlassian.net/browse/AAQ-348?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: core_backend/migrations/versions/f269c75dbf69_add_content_table.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    op.create_table(
+        "content",
+        sa.Column("content_id", sa.Integer(), nullable=False),
+        sa.Column(
+            "content_embedding",
+            pgvector.sqlalchemy.Vector(dim=int(PGVECTOR_VECTOR_SIZE)),
+            nullable=False,
+        ),
+        sa.Column("content_title", sa.String(length=150), nullable=False),
+        sa.Column("content_text", sa.String(length=14460), nullable=False),
```
Comment (lickem22) [inline core_backend/migrations/versions/f269c75dbf69_add_content_table.py:40]: @suzinyou , we need to find a bigger number for the content_text. 1446 made the validatinon script fail. So I increased the number.

Comment (suzinyou) [reply core_backend/migrations/versions/f269c75dbf69_add_content_table.py:40]: Ah, and the validation script doesn't use the ORM to add data?

File: core_backend/tests/api/test.env [hunk 2]
```diff
@@ -1,10 +1,6 @@
 POSTGRES_USER=postgres-test-user
 POSTGRES_PASSWORD=postgres-test-pw
 POSTGRES_DB=postgres-test-db
-POSTGRES_PORT=5433
```
Comment (suzinyou) [inline core_backend/tests/api/test.env:4]: Why did you delete this? By default POSTGRES_PORT=5432 in app_config.py. However, in the make setup-test-db target we launch a container to be exposed to port 5433!

Comment (suzinyou) [reply core_backend/tests/api/test.env:4]: Running `make tests` in core_backend is failing due to this line being removed.

File: core_backend/validation/validate_retrieval.py [hunk 3]
```diff
@@ -104,43 +98,40 @@ def prepare_content_data(
...
         )
         content_embeddings = [x["embedding"] for x in embedding_results.data]
 
-        points = [
-            PointStruct(
-                id=str(content_id),
-                vector=content_embedding,
-                payload=payload,
+        contents = [
+            ContentDB(
```
Comment (suzinyou) [inline core_backend/validation/validate_retrieval.py:115]: Wait, I see that we are using ContentDB to add the contents here, but there the text limit is set to 1446. So it doesn't seem like we hit the character limit with the validation dataset?

File: core_backend/app/db/db_models.py [hunk 4]
```diff
@@ -294,3 +309,196 @@ async def save_waresponse_to_db(
...
+    """
+
+    __tablename__ = "content"
+
+    content_id: Mapped[int] = mapped_column(Integer, primary_key=True, nullable=False)
+    content_embedding: Mapped[Vector] = mapped_column(
+        Vector(int(PGVECTOR_VECTOR_SIZE)), nullable=False
+    )
+    content_title: Mapped[str] = mapped_column(String(length=150), nullable=False)
+    content_text: Mapped[str] = mapped_column(String(length=1446), nullable=False)
```
Comment (suzinyou) [inline core_backend/app/db/db_models.py:326]: Here the length is at 1446

File: core_backend/deployment/bootstrap_demo.sh
Comment (suzinyou) [inline core_backend/deployment/bootstrap_demo.sh:1]: FYI these files are incomplete and will be modified by AAQ-278 PR!

File: core_backend/migrations/versions/f269c75dbf69_add_content_table.py (hunk 1 above)
Comment (suzinyou) [reply core_backend/migrations/versions/f269c75dbf69_add_content_table.py:40]: Could you point me to the workflow log where this failure happened?

Comment (lickem22) [reply core_backend/migrations/versions/f269c75dbf69_add_content_table.py:40]: No it directly adds it using the session. But it doesn't really matter as the max length would still be too low.

**Review by suzinyou — CHANGES_REQUESTED** [review]
Thanks so much! Please check out my comments! And for the text length limit I don't think I fully understand what's going on, we can discuss

File: core_backend/app/db/db_models.py (hunk 4 above)
Comment (lickem22) [reply core_backend/app/db/db_models.py:326]: Yes but ContentDB doesn't do any check. It is just a reference to get the migrations. Meaning that the constraint is on the DB level not on the code level. So if we add data to the DB using ContentDB. The code won't check the character length but the query will fail on postgres

File: core_backend/migrations/versions/f269c75dbf69_add_content_table.py (hunk 1 above)
Comment (lickem22) [reply core_backend/migrations/versions/f269c75dbf69_add_content_table.py:40]: Here is the [link](https://github.com/IDinsight/aaq-core/actions/runs/8064419670/job/22028262652)

File: docs/images/architecture-docker.png
[the remark this answers is not in the document]
Comment (lickem22) [reply docs/images/architecture-docker.png:1]: What do you mean by the white margins are even?

File: core_backend/tests/api/test.env (hunk 2 above)
Comment (lickem22) [reply core_backend/tests/api/test.env:4]: Will fix it

File: core_backend/app/db/db_models.py (hunk 4 above)
Comment (lickem22) [reply core_backend/app/db/db_models.py:326]: The best way to address that issue is to set the max character to a higher value. Is the current value a hard limit? Otherwise we can limit the text of each FAQ to be 1446 in the validation script.

Comment (lickem22) [reply core_backend/app/db/db_models.py:326]: I will increase the limit to 2000. We need to figure out how to handle communicating with whatsapp because I understand whatsapp have a max character limit

File: infrastructure/demo/main/ecs.tf
Comment (suzinyou) [inline infrastructure/demo/main/ecs.tf:1]: Can we undo this change? Otherwise testing and modifying CI/CD pipelines will get complicated.. And I don't think we should wait till CI/CD gets merged for this to get merged!

File: infrastructure/demo/main/iam.tf
Comment (suzinyou) [inline infrastructure/demo/main/iam.tf:1]: Same here, let's remove this in another PR

**Review by suzinyou — CHANGES_REQUESTED** [review]
Hey @lickem22 let a few more comments!

File: docs/images/architecture-docker.png
[the remark this answers is not in the document]
Comment (suzinyou) [reply docs/images/architecture-docker.png:1]: The image had different margins on left and right. Now it looks even so all good

File: infrastructure/demo/main/ecs.tf
Comment (lickem22) [reply infrastructure/demo/main/ecs.tf:1]: It's done. I also updated validation_retrival like you said.

File: infrastructure/demo/main/iam.tf [hunk 5]
```diff
@@ -85,6 +85,7 @@ data "aws_iam_policy_document" "web_ec2_role_policy" {
     resources = [
       aws_service_discovery_service.backend.arn,
       aws_service_discovery_service.vectordb.arn,
+      aws_service_discovery_service.backend.arn
```
Comment (suzinyou) [inline infrastructure/demo/main/iam.tf:88]: Seems like you accidentally added a duplicate!
```suggestion
```

**Review by suzinyou — APPROVED** [review]
Thanks SO much Carlos!! ~Can you commit the suggestion I made and merge it?~ I made the quick change :)
